[BUG] terramate run terraform init etc. crashes

**Describe the bug**
'terramate generate' works great, and creates the tf files in a stack. And these files work with direct terraform invocation (ie. terraform init, terraform plan etc.) But if I try 'terramate run terraform init' etc. I get a runtime error:

erramate run terraform plan
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x13a43db]

goroutine 1 [running]:
github.com/mineiros-io/terramate/git.(*Git).Remotes(0x0)
	github.com/mineiros-io/terramate/git/git.go:325 +0x21b
github.com/mineiros-io/terramate/cmd/terramate/cli.project.checkDefaultRemote({{0xc0000ae700, 0x44}, {0xc0000ae700, 0x6a}, 0x0, {{0xc0000ae700, 0x44}, 0xc0001e0df8, 0x0}, {0x0, ...}, ...})
	github.com/mineiros-io/terramate/cmd/terramate/cli/project.go:184 +0x292
github.com/mineiros-io/terramate/cmd/terramate/cli.(*cli).checkGit(0xc000292b60)
	github.com/mineiros-io/terramate/cmd/terramate/cli/cli.go:363 +0x2d9
github.com/mineiros-io/terramate/cmd/terramate/cli.(*cli).runOnStacks(0xc000292b60)
	github.com/mineiros-io/terramate/cmd/terramate/cli/cli.go:878 +0x2f0
github.com/mineiros-io/terramate/cmd/terramate/cli.(*cli).run(0xc000292b60)
	github.com/mineiros-io/terramate/cmd/terramate/cli/cli.go:336 +0x514
github.com/mineiros-io/terramate/cmd/terramate/cli.Exec({0xc0000b0010, 0x3, 0x3}, {0x15abee0, 0xc0000a2000}, {0x15abf00, 0xc0000a2008}, {0x15abf00, 0xc0000a2010})
	github.com/mineiros-io/terramate/cmd/terramate/cli/cli.go:139 +0xb7
main.main()
	github.com/mineiros-io/terramate/cmd/terramate/main.go:24 +0x7a
**To Reproduce**

I am calling a module via a source of a relative path that leads to a directory outside of the terramate root, so maybe that is a
factor


Steps to reproduce the behavior:
Just running 'terramate run terraform <cmd>' with terraform version 1.1.7 (or 1.2.2) and terramate 0.1.5

**Expected behavior**
Expect the terraform command to be executed

**Log Output**
Add logs from Terramate to help debug your problem.

**Environment (please complete the following information):**
 - OS: OSX
 - OS Version Big Sur 11.6.5
 - Git Version N/A
 -  - Terramate Version 0.1.5

**Additional context**
Add any other context about the problem here.

--- a/cmd/terramate/cli/cli.go
+++ b/cmd/terramate/cli/cli.go
@@ -358,6 +358,10 @@ func (c *cli) checkGit() {
 		Str("action", "checkGit()").
 		Logger()
 
+	if !c.prj.isRepo {
+		return
+	}
+
 	logger.Trace().Msg("Check git default remote.")
 
 	if err := c.prj.checkDefaultRemote(); err != nil {

--- a/cmd/terramate/e2etests/exp_globals_test.go
+++ b/cmd/terramate/e2etests/exp_globals_test.go
@@ -229,16 +229,25 @@ stack "/stacks/stack-2":
 
 	for _, tcase := range tcases {
 		t.Run(tcase.name, func(t *testing.T) {
-			s := sandbox.New(t)
-			s.BuildTree(tcase.layout)
-
-			for _, globalBlock := range tcase.globals {
-				path := filepath.Join(s.RootDir(), globalBlock.path)
-				test.AppendFile(t, path, config.DefaultFilename, globalBlock.add.String())
+			sandboxes := []sandbox.S{
+				sandbox.New(t),
+				sandbox.NoGit(t),
 			}
 
-			ts := newCLI(t, project.AbsPath(s.RootDir(), tcase.wd))
-			assertRunResult(t, ts.run("experimental", "globals"), tcase.want)
+			for _, s := range sandboxes {
+				s.BuildTree(tcase.layout)
+
+				for _, globalBlock := range tcase.globals {
+					path := filepath.Join(s.RootDir(), globalBlock.path)
+					test.AppendFile(t, path, config.DefaultFilename,
+						globalBlock.add.String())
+				}
+
+				test.WriteRootConfig(t, s.RootDir())
+
+				ts := newCLI(t, project.AbsPath(s.RootDir(), tcase.wd))
+				assertRunResult(t, ts.run("experimental", "globals"), tcase.want)
+			}
 		})
 	}
 }

--- a/cmd/terramate/e2etests/list_git_test.go
+++ b/cmd/terramate/e2etests/list_git_test.go
@@ -1,0 +1,193 @@
+package e2etest
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/mineiros-io/terramate/config"
+	"github.com/mineiros-io/terramate/test"
+	"github.com/mineiros-io/terramate/test/sandbox"
+)
+
+func TestE2EListWithGit(t *testing.T) {
+	for _, tc := range listTestcases() {
+		t.Run(tc.name, func(t *testing.T) {
+			s := sandbox.New(t)
+			s.BuildTree(tc.layout)
+
+			cli := newCLI(t, s.RootDir())
+			assertRunResult(t, cli.listStacks(), tc.want)
+		})
+	}
+}
+
+func TestListDetectChangesInSubDirOfStack(t *testing.T) {
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack")
+	subdir := stack.CreateDir("sub/dir")
+	subfile := subdir.CreateFile("something.sh", "# nothing")
+
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+	git.CheckoutNew("change-the-stack")
+
+	subfile.Write("# changed")
+	git.Add(stack.Path())
+	git.Commit("stack changed")
+
+	want := runExpected{
+		Stdout: stack.RelPath() + "\n",
+	}
+	assertRunResult(t, cli.listChangedStacks(), want)
+}
+
+func TestListDetectChangesInSubDirOfStackWithOtherConfigs(t *testing.T) {
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack")
+	subdir := stack.CreateDir("sub")
+	subsubdir := subdir.CreateDir("dir")
+	subsubfile := subsubdir.CreateFile("something.sh", "# nothing")
+
+	subdir.CreateFile(config.DefaultFilename, `
+terramate {
+	
+}	
+`)
+
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+	git.CheckoutNew("change-the-stack")
+
+	subsubfile.Write("# changed")
+	git.Add(stack.Path())
+	git.Commit("stack changed")
+
+	want := runExpected{
+		Stdout: stack.RelPath() + "\n",
+	}
+	assertRunResult(t, cli.listChangedStacks(), want)
+}
+
+func TestListChangedIgnoreDeletedStackDirectory(t *testing.T) {
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack-old")
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+	git.CheckoutNew("deleted-stack")
+
+	test.RemoveAll(t, stack.Path())
+
+	git.CommitAll("removed stack")
+
+	assertRun(t, cli.listChangedStacks())
+}
+
+func TestListChangedIgnoreDeletedNonStackDirectory(t *testing.T) {
+	s := sandbox.New(t)
+
+	s.CreateStack("stack")
+	toBeDeletedDir := filepath.Join(s.RootDir(), "to-be-deleted")
+	test.MkdirAll(t, toBeDeletedDir)
+	test.WriteFile(t, toBeDeletedDir, "test.txt", "")
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+
+	git.CheckoutNew("deleted-diretory")
+
+	test.RemoveAll(t, toBeDeletedDir)
+	git.CommitAll("removed directory")
+
+	assertRun(t, cli.listChangedStacks())
+}
+
+func TestListChangedDontIgnoreStackDeletedFiles(t *testing.T) {
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack")
+	testDir := stack.CreateDir("test")
+	file := testDir.CreateFile("testfile", "")
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+	git.CheckoutNew("deleted-file")
+
+	test.RemoveAll(t, file.Path())
+
+	git.CommitAll("removed file")
+
+	assertRunResult(t, cli.listChangedStacks(), runExpected{
+		Stdout: stack.RelPath() + "\n",
+	})
+}
+
+func TestListChangedDontIgnoreStackDeletedDirectories(t *testing.T) {
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack")
+	testDir := stack.CreateDir("test")
+	testDir.CreateFile("testfile1", "")
+	testDir.CreateFile("testfile2", "")
+	cli := newCLI(t, s.RootDir())
+
+	git := s.Git()
+	git.CommitAll("all")
+	git.Push("main")
+	git.CheckoutNew("deleted-dir")
+
+	test.RemoveAll(t, testDir.Path())
+
+	git.CommitAll("removed directory")
+
+	assertRunResult(t, cli.listChangedStacks(), runExpected{
+		Stdout: stack.RelPath() + "\n",
+	})
+}
+
+func TestListTwiceBug(t *testing.T) {
+	const (
+		mainTfFileName = "main.tf"
+		modname        = "modA"
+	)
+
+	s := sandbox.New(t)
+
+	stack := s.CreateStack("stack")
+	mod1 := s.CreateModule(modname)
+	mod1MainTf := mod1.CreateFile(mainTfFileName, "# module A")
+
+	stack.CreateFile("main.tf", `
+module "mod1" {
+source = "%s"
+}`, stack.ModSource(mod1))
+
+	git := s.Git()
+	git.CommitAll("first commit")
+	git.Push("main")
+	git.CheckoutNew("change-stack")
+
+	mod1MainTf.Write("# something else")
+	stack.CreateFile("test.txt", "something else")
+	git.CommitAll("stack and module changed")
+
+	cli := newCLI(t, s.RootDir())
+
+	wantList := stack.RelPath() + "\n"
+	assertRunResult(t, cli.listChangedStacks(), runExpected{Stdout: wantList})
+}

--- a/cmd/terramate/e2etests/list_git_test.go
+++ b/cmd/terramate/e2etests/list_git_test.go
@@ -1,3 +1,17 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package e2etest
 
 import (

--- a/cmd/terramate/e2etests/list_nongit_test.go
+++ b/cmd/terramate/e2etests/list_nongit_test.go
@@ -1,0 +1,22 @@
+package e2etest
+
+import (
+	"testing"
+
+	"github.com/mineiros-io/terramate/test"
+	"github.com/mineiros-io/terramate/test/sandbox"
+)
+
+func TestE2EListNonGit(t *testing.T) {
+	for _, tc := range listTestcases() {
+		t.Run(tc.name, func(t *testing.T) {
+			s := sandbox.NoGit(t)
+			s.BuildTree(tc.layout)
+
+			test.WriteRootConfig(t, s.RootDir())
+
+			cli := newCLI(t, s.RootDir())
+			assertRunResult(t, cli.listStacks(), tc.want)
+		})
+	}
+}

--- a/cmd/terramate/e2etests/list_nongit_test.go
+++ b/cmd/terramate/e2etests/list_nongit_test.go
@@ -1,3 +1,17 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package e2etest
 
 import (

--- a/cmd/terramate/e2etests/list_test.go
+++ b/cmd/terramate/e2etests/list_test.go
@@ -15,23 +15,21 @@
 package e2etest
 
 import (
-	"path/filepath"
 	"testing"
 
-	"github.com/mineiros-io/terramate/config"
 	"github.com/mineiros-io/terramate/hcl"
 	"github.com/mineiros-io/terramate/test"
 	"github.com/mineiros-io/terramate/test/sandbox"
 )
 
-func TestCLIList(t *testing.T) {
-	type testcase struct {
-		name   string
-		layout []string
-		want   runExpected
-	}
+type testcase struct {
+	name   string
+	layout []string
+	want   runExpected
+}
 
-	for _, tc := range []testcase{
+func listTestcases() []testcase {
+	return []testcase{
 		{
 			name: "no stack",
 		},
@@ -118,14 +116,6 @@ z/a
 `,
 			},
 		},
-	} {
-		t.Run(tc.name, func(t *testing.T) {
-			s := sandbox.New(t)
-			s.BuildTree(tc.layout)
-
-			cli := newCLI(t, s.RootDir())
-			assertRunResult(t, cli.listStacks(), tc.want)
-		})
 	}
 }
 
@@ -159,175 +149,4 @@ func TestListNoSuchFile(t *testing.T) {
 		Status:      1,
 		StderrRegex: "no such file or directory",
 	})
-}
-
-func TestListDetectChangesInSubDirOfStack(t *testing.T) {
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack")
-	subdir := stack.CreateDir("sub/dir")
-	subfile := subdir.CreateFile("something.sh", "# nothing")
-
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-	git.CheckoutNew("change-the-stack")
-
-	subfile.Write("# changed")
-	git.Add(stack.Path())
-	git.Commit("stack changed")
-
-	want := runExpected{
-		Stdout: stack.RelPath() + "\n",
-	}
-	assertRunResult(t, cli.listChangedStacks(), want)
-}
-
-func TestListDetectChangesInSubDirOfStackWithOtherConfigs(t *testing.T) {
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack")
-	subdir := stack.CreateDir("sub")
-	subsubdir := subdir.CreateDir("dir")
-	subsubfile := subsubdir.CreateFile("something.sh", "# nothing")
-
-	subdir.CreateFile(config.DefaultFilename, `
-terramate {
-	
-}	
-`)
-
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-	git.CheckoutNew("change-the-stack")
-
-	subsubfile.Write("# changed")
-	git.Add(stack.Path())
-	git.Commit("stack changed")
-
-	want := runExpected{
-		Stdout: stack.RelPath() + "\n",
-	}
-	assertRunResult(t, cli.listChangedStacks(), want)
-}
-
-func TestListChangedIgnoreDeletedStackDirectory(t *testing.T) {
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack-old")
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-	git.CheckoutNew("deleted-stack")
-
-	test.RemoveAll(t, stack.Path())
-
-	git.CommitAll("removed stack")
-
-	assertRun(t, cli.listChangedStacks())
-}
-
-func TestListChangedIgnoreDeletedNonStackDirectory(t *testing.T) {
-	s := sandbox.New(t)
-
-	s.CreateStack("stack")
-	toBeDeletedDir := filepath.Join(s.RootDir(), "to-be-deleted")
-	test.MkdirAll(t, toBeDeletedDir)
-	test.WriteFile(t, toBeDeletedDir, "test.txt", "")
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-
-	git.CheckoutNew("deleted-diretory")
-
-	test.RemoveAll(t, toBeDeletedDir)
-	git.CommitAll("removed directory")
-
-	assertRun(t, cli.listChangedStacks())
-}
-
-func TestListChangedDontIgnoreStackDeletedFiles(t *testing.T) {
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack")
-	testDir := stack.CreateDir("test")
-	file := testDir.CreateFile("testfile", "")
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-	git.CheckoutNew("deleted-file")
-
-	test.RemoveAll(t, file.Path())
-
-	git.CommitAll("removed file")
-
-	assertRunResult(t, cli.listChangedStacks(), runExpected{
-		Stdout: stack.RelPath() + "\n",
-	})
-}
-
-func TestListChangedDontIgnoreStackDeletedDirectories(t *testing.T) {
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack")
-	testDir := stack.CreateDir("test")
-	testDir.CreateFile("testfile1", "")
-	testDir.CreateFile("testfile2", "")
-	cli := newCLI(t, s.RootDir())
-
-	git := s.Git()
-	git.CommitAll("all")
-	git.Push("main")
-	git.CheckoutNew("deleted-dir")
-
-	test.RemoveAll(t, testDir.Path())
-
-	git.CommitAll("removed directory")
-
-	assertRunResult(t, cli.listChangedStacks(), runExpected{
-		Stdout: stack.RelPath() + "\n",
-	})
-}
-
-func TestListTwiceBug(t *testing.T) {
-	const (
-		mainTfFileName = "main.tf"
-		modname        = "modA"
-	)
-
-	s := sandbox.New(t)
-
-	stack := s.CreateStack("stack")
-	mod1 := s.CreateModule(modname)
-	mod1MainTf := mod1.CreateFile(mainTfFileName, "# module A")
-
-	stack.CreateFile("main.tf", `
-module "mod1" {
-source = "%s"
-}`, stack.ModSource(mod1))
-
-	git := s.Git()
-	git.CommitAll("first commit")
-	git.Push("main")
-	git.CheckoutNew("change-stack")
-
-	mod1MainTf.Write("# something else")
-	stack.CreateFile("test.txt", "something else")
-	git.CommitAll("stack and module changed")
-
-	cli := newCLI(t, s.RootDir())
-
-	wantList := stack.RelPath() + "\n"
-	assertRunResult(t, cli.listChangedStacks(), runExpected{Stdout: wantList})
 }

--- a/cmd/terramate/e2etests/run_test.go
+++ b/cmd/terramate/e2etests/run_test.go
@@ -453,16 +453,23 @@ stack-a
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
-			s := sandbox.New(t)
-			s.BuildTree(tc.layout)
-
-			wd := s.RootDir()
-			if tc.workingDir != "" {
-				wd = filepath.Join(wd, tc.workingDir)
+			sandboxes := []sandbox.S{
+				sandbox.New(t),
+				sandbox.NoGit(t),
 			}
 
-			cli := newCLI(t, wd)
-			assertRunResult(t, cli.stacksRunOrder(), tc.want)
+			for _, s := range sandboxes {
+				s.BuildTree(tc.layout)
+				test.WriteRootConfig(t, s.RootDir())
+
+				wd := s.RootDir()
+				if tc.workingDir != "" {
+					wd = filepath.Join(wd, tc.workingDir)
+				}
+
+				cli := newCLI(t, wd)
+				assertRunResult(t, cli.stacksRunOrder(), tc.want)
+			}
 		})
 	}
 }
@@ -632,18 +639,27 @@ func TestRunWants(t *testing.T) {
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
-			s := sandbox.New(t)
-			s.BuildTree(tc.layout)
+			sandboxes := []sandbox.S{
+				sandbox.New(t),
+				sandbox.NoGit(t),
+			}
 
-			cli := newCLI(t, filepath.Join(s.RootDir(), tc.wd))
-			assertRunResult(t, cli.stacksRunOrder(), tc.want)
+			for _, s := range sandboxes {
+				s.BuildTree(tc.layout)
+				test.WriteRootConfig(t, s.RootDir())
 
-			// required because `terramate run` requires a clean repo.
-			git := s.Git()
-			git.CommitAll("everything")
+				cli := newCLI(t, filepath.Join(s.RootDir(), tc.wd))
+				assertRunResult(t, cli.stacksRunOrder(), tc.want)
 
-			// TODO(i4k): not portable
-			assertRunResult(t, cli.run("run", "sh", "-c", "pwd | xargs basename"), tc.want)
+				if s.IsGit() {
+					// required because `terramate run` requires a clean repo.
+					git := s.Git()
+					git.CommitAll("everything")
+				}
+
+				// TODO(i4k): not portable
+				assertRunResult(t, cli.run("run", "sh", "-c", "pwd | xargs basename"), tc.want)
+			}
 		})
 	}
 }

--- a/test/hcl.go
+++ b/test/hcl.go
@@ -93,3 +93,14 @@ func assertStackBlock(t *testing.T, got, want *hcl.Stack) {
 		assert.EqualStrings(t, w, got.After[i], "stack after mismatch")
 	}
 }
+
+// WriteRootConfig writes a basic terramate root config.
+func WriteRootConfig(t *testing.T, rootdir string) {
+	WriteFile(t, rootdir, "root.config.tm", `
+terramate {
+	config {
+
+	}
+}
+			`)
+}

--- a/test/sandbox/sandbox.go
+++ b/test/sandbox/sandbox.go
@@ -106,10 +106,21 @@ func NewWithGitConfig(t *testing.T, cfg GitConfig) S {
 func NoGit(t *testing.T) S {
 	t.Helper()
 
-	rootdir := test.CanonPath(t, t.TempDir())
+	// here we create some stacks outside the root directory of the
+	// sandbox so we can check if terramate does not ascend to parent
+	// directories.
+
+	outerDir := t.TempDir()
+	buildTree(t, outerDir, []string{
+		"s:this-stack-must-never-be-visible",
+		"s:other-hidden-stack",
+	})
+	rootdir := filepath.Join(outerDir, "sandbox")
+	test.MkdirAll(t, rootdir)
+
 	return S{
 		t:       t,
-		rootdir: rootdir,
+		rootdir: test.CanonPath(t, rootdir),
 	}
 }
 
@@ -136,78 +147,11 @@ func NoGit(t *testing.T) S {
 // This is an internal mini-lang used to simplify testcases, so it expects well
 // formed layout specification.
 func (s S) BuildTree(layout []string) {
-	t := s.t
-	t.Helper()
-
-	parsePathData := func(spec string) (string, string) {
-		tmp := spec[2:]
-		if len(tmp) == 0 {
-			// relative to s.rootdir
-			return ".", ""
-		}
-		index := strings.IndexByte(tmp, ':')
-		if index == -1 {
-			return tmp, ""
-		}
-		path := tmp[0:index]
-		data := tmp[index+1:]
-		return path, data
-	}
-
-	gentmfile := func(relpath, data string) {
-		attrs := strings.Split(data, ";")
-
-		cfgdir := filepath.Join(s.RootDir(), relpath)
-		test.MkdirAll(t, cfgdir)
-		cfg, err := hcl.NewConfig(cfgdir)
-		assert.NoError(t, err)
-
-		cfg.Stack = &hcl.Stack{}
-		cfg.Terramate = &hcl.Terramate{}
-
-		for _, attr := range attrs {
-			parts := strings.Split(attr, "=")
-			name := parts[0]
-			value := parts[1]
-			switch name {
-			case "after":
-				cfg.Stack.After = parseListSpec(t, name, value)
-			case "before":
-				cfg.Stack.Before = parseListSpec(t, name, value)
-			case "wants":
-				cfg.Stack.Wants = parseListSpec(t, name, value)
-			case "description":
-				cfg.Stack.Description = value
-			default:
-				t.Fatalf("attribute " + parts[0] + " not supported.")
-			}
-		}
-
-		assert.NoError(t, cfg.Save(config.DefaultFilename),
-			"BuildTree() failed to generate config file.")
-	}
-
-	for _, spec := range layout {
-		path, data := parsePathData(spec)
-
-		specKind := string(spec[0:2])
-		switch specKind {
-		case "d:":
-			test.MkdirAll(t, filepath.Join(s.rootdir, spec[2:]))
-		case "s:":
-			if data == "" {
-				s.CreateStack(path)
-				continue
-			}
-
-			gentmfile(path, data)
-		case "f:":
-			test.WriteFile(t, s.rootdir, path, data)
-		default:
-			t.Fatalf("unknown spec kind: %q", specKind)
-		}
-	}
+	buildTree(s.t, s.RootDir(), layout)
 }
+
+// IsGit tells if the sandbox is a git repository.
+func (s S) IsGit() bool { return s.git != nil }
 
 // Git returns a git wrapper that is useful to run git commands safely inside
 // the test env repo.
@@ -522,4 +466,79 @@ func parseListSpec(t *testing.T, name, value string) []string {
 	}
 
 	return list
+}
+
+func buildTree(t *testing.T, rootdir string, layout []string) {
+	t.Helper()
+
+	parsePathData := func(spec string) (string, string) {
+		tmp := spec[2:]
+		if len(tmp) == 0 {
+			// relative to s.rootdir
+			return ".", ""
+		}
+		index := strings.IndexByte(tmp, ':')
+		if index == -1 {
+			return tmp, ""
+		}
+		path := tmp[0:index]
+		data := tmp[index+1:]
+		return path, data
+	}
+
+	gentmfile := func(relpath, data string) {
+		attrs := strings.Split(data, ";")
+
+		cfgdir := filepath.Join(rootdir, relpath)
+		test.MkdirAll(t, cfgdir)
+		cfg, err := hcl.NewConfig(cfgdir)
+		assert.NoError(t, err)
+
+		cfg.Stack = &hcl.Stack{}
+		cfg.Terramate = &hcl.Terramate{}
+
+		for _, attr := range attrs {
+			parts := strings.Split(attr, "=")
+			name := parts[0]
+			value := parts[1]
+			switch name {
+			case "after":
+				cfg.Stack.After = parseListSpec(t, name, value)
+			case "before":
+				cfg.Stack.Before = parseListSpec(t, name, value)
+			case "wants":
+				cfg.Stack.Wants = parseListSpec(t, name, value)
+			case "description":
+				cfg.Stack.Description = value
+			default:
+				t.Fatalf("attribute " + parts[0] + " not supported.")
+			}
+		}
+
+		assert.NoError(t, cfg.Save(config.DefaultFilename),
+			"BuildTree() failed to generate config file.")
+	}
+
+	for _, spec := range layout {
+		path, data := parsePathData(spec)
+
+		specKind := string(spec[0:2])
+		switch specKind {
+		case "d:":
+			test.MkdirAll(t, filepath.Join(rootdir, spec[2:]))
+		case "s:":
+			if data == "" {
+				abspath := filepath.Join(rootdir, path)
+				test.MkdirAll(t, abspath)
+				assert.NoError(t, terramate.Init(rootdir, abspath))
+				continue
+			}
+
+			gentmfile(path, data)
+		case "f:":
+			test.WriteFile(t, rootdir, path, data)
+		default:
+			t.Fatalf("unknown spec kind: %q", specKind)
+		}
+	}
 }

